### PR TITLE
ApiTestCase and assertions docs

### PR DESCRIPTION
**Pull request by dunglas**
Docs for:

* https://github.com/api-platform/core/pull/2608
* https://github.com/api-platform/core/pull/2887
* https://github.com/api-platform/core/pull/3003
* https://github.com/api-platform/core/pull/2940

See also the PR on the demo app: https://github.com/api-platform/demo/pull/129

### CODE REVIEW

File: distribution/testing.md [hunk 1]
```diff
@@ -248,17 +206,22 @@ transaction previously begun. Because of this, you can run your tests without wo
...
-and tested thanks to [PHPUnit](https://phpunit.de/index.html)!
+Check out the [testing documentation](../core/testing.md) to discover the full range of assertions and other features provided by API Platform's test utilities.
 
-### Additional and Alternative Testing Tools
+## Writing Unit Tests
+
+In addition to integration tests written using the helpers provided by `ApiTestCase`, all the classes of your project should be covered by [unit tests](https://en.wikipedia.org/wiki/Unit_testing).
+To do so, learn how to write unit tests with [PHPUnit](https://phpunit.de/index.html) and [its Symfony/API Platform integration](https://symfony.com/doc/current/testing.html).
+
+## Additional and Alternative Testing Tools
```
Comment (dunglas) [inline distribution/testing.md]: Maybe should we drop entirely this section. I'm not sure it's still relevant/wanted.

File: core/testing.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
+# Testing Utilities
+
+API Platform Core provides a set of useful utilities dedicated to API testing.
+For an overview of how to test an API Platform app, be sure to read [the testing cookbook first](../distribution/testing.md).
+
+## The Test HttpClient
+
+API Platform provides its own implementation of the [Symfony HttpClient](https://symfony.com/doc/current/components/http_client.html)'s interfaces, tailored to be used directly in [PHPUnit](https://phpunit.de/) test classes.
+
+While all the convenient features of Symfony HttpClient are available and usable directly, under the hood the API Platform implementation manipulates directly [the Symfony HttpKernel](https://symfony.com/doc/current/components/http_kernel.html) to simulate HTTP requests and responses.
```
Comment (bendavies) [inline core/testing.md]:
```suggestion
While all the convenient features of Symfony HttpClient are available and usable directly, under the hood the API Platform implementation manipulates [the Symfony HttpKernel](https://symfony.com/doc/current/components/http_kernel.html) directly to simulate HTTP requests and responses.
```

File: core/testing.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+This approach allows a huge boost performance compared to triggering real network requests.
```
Comment (bendavies) [inline core/testing.md]:
```suggestion
This approach results in a huge performance boost compared to triggering real network requests.
```

File: core/testing.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+It also allows to access to the Symfony kernel, to the [Dependency Injection Container](https://symfony.com/doc/current/testing.html#accessing-the-container) and so to all your services!
```
Comment (bendavies) [inline core/testing.md]:
```suggestion
It also allows to access to the [Symfony HttpKernel](https://symfony.com/doc/current/components/http_kernel.html and to all your services via the [Dependency Injection Container](https://symfony.com/doc/current/testing.html#accessing-the-container) 
```

File: core/testing.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+Reuse them to run - for instance - SQL queries or requests to external APIs directly from your tests.
```
Comment (bendavies) [inline core/testing.md]:
```suggestion
Reuse them to run, for instance, SQL queries or requests to external APIs directly from your tests.
```

File: core/testing.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        $this->assertJsonContains(/* a JSON document as an array or as a string */);
+
+        // justinrainbow/json-schema must be installed to use the following assertions
+
+        // Asserts that the returned JSON matches the passed JSON Schema
+        $this->assertMatchesJsonSchema(/* a JSON Schema as an array or as a string */);
+
+        // Asserts that the returned JSON is validated by the JSON Schema generated for this resource by API Platform
+
+        // For collections (like)
```
Comment (bendavies) [inline core/testing.md]:
```suggestion
        // For collections
```

File: distribution/testing.md [hunk 7]
```diff
@@ -52,191 +53,148 @@ You can now load your fixtures in the database with the following command:
...
 use App\Entity\Book;
 use Hautelook\AliceBundle\PhpUnit\RefreshDatabaseTrait;
-use Symfony\Bundle\FrameworkBundle\Client;
-use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
-use Symfony\Component\HttpFoundation\Response;
 
-class ApiTest extends WebTestCase
+class BooksTest extends ApiTestCase
 {
+    // This trait provided by HautelookAliceBundle will take care of refreshing the database content to put it in a known state between every tests
```
Comment (bendavies) [inline distribution/testing.md]:
```suggestion
    // This trait provided by HautelookAliceBundle will take care of refreshing the database content to a known state before each test
```

**Comment by dunglas** [comment]
Thanks for the review @bendavies!!

**Pull request merged**